promql/engine.go

```diff
@@ -2825,6 +2825,44 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
 	panic(fmt.Errorf("operator %q not allowed for Scalar operations", op))
 }
 
+// Helper function to trim native histogram buckets.
+func trimBuckets(trimmedHist, hlhs *histogram.FloatHistogram, rhs float64, mode string) {
+	trimmedHist.PositiveBuckets = make([]float64, len(hlhs.PositiveBuckets))
+	copy(trimmedHist.PositiveBuckets, hlhs.PositiveBuckets)
+
+	trimmedHist.NegativeBuckets = make([]float64, len(hlhs.NegativeBuckets))
+	copy(trimmedHist.NegativeBuckets, hlhs.NegativeBuckets)
+
+	updatedCount := 0.0
+	updatedSum := 0.0
+
+	for i, iter := 0, hlhs.PositiveBucketIterator(); iter.Next(); i++ {
+		bucket := iter.At()
+		if (mode == "LTRIM" && bucket.Upper > rhs) || (mode == "RTRIM" && bucket.Lower < rhs) {
+			trimmedHist.PositiveBuckets[i] = 0
+		} else {
+			updatedCount += bucket.Count
+			bucketMidpoint := (bucket.Lower + bucket.Upper) / 2
+			updatedSum += bucket.Count * bucketMidpoint
+		}
+	}
+
+	for i, iter := 0, hlhs.NegativeBucketIterator(); iter.Next(); i++ {
```

Comment:
Instead of iterating through positive and negative buckets separately, we could simplify the code by using `AllBucketIterator` to iterate over both at once. This might make the logic cleaner and easier to maintain.